cimbar/cimbar.py

```diff
@@ -135,31 +259,123 @@ def update(c, r, g, b):
     else:
         pos = [(67, 0), (0, 67), (conf.TOTAL_SIZE-79, 0), (0, conf.TOTAL_SIZE-79)]
 
+    colors = []
     for x, y in pos:
         iblock = img.crop((x, y, x + 4, y + 4))
-        r, g, b = avg_color(iblock)
-        update(cc, *avg_color(iblock))
+        ac = avg_color(iblock, False)
+        colors.append(ac)
+        update(cc, *ac)
 
     print(f'tint is {cc}')
     return cc['r'], cc['g'], cc['b']
 
 
-def _decode_iter(ct, img, color_img):
+def _crop_cell(img, x, y):
+    return img.crop((x+1, y+1, x + conf.CELL_SIZE-1, y + conf.CELL_SIZE-1))
+
+
+def _decode_symbols(ct, img):
     cell_pos, num_edge_cells = cell_positions(conf.CELL_SPACING_X, conf.CELL_SPACING_Y, conf.CELL_DIM_X,
                                               conf.CELL_DIM_Y, conf.CELLS_OFFSET, conf.MARKER_SIZE_X, conf.MARKER_SIZE_Y)
     finder = AdjacentCellFinder(cell_pos, num_edge_cells, conf.CELL_DIM_X, conf.MARKER_SIZE_X)
     decode_order = FloodDecodeOrder(cell_pos, finder)
+    print('beginning decode symbols pass...')
     for i, (x, y), drift in decode_order:
-        best_bits, best_dx, best_dy, best_distance = _decode_cell(ct, img, color_img, x, y, drift)
+        best_bits, best_cell, best_dx, best_dy, best_distance = _decode_cell(ct, img, x, y, drift)
         decode_order.update(best_dx, best_dy, best_distance)
-        yield i, best_bits
+        yield i, best_bits, best_cell
+
+
+def _calc_ccm(ct, color_lookups, cc_setting, state_info):
+    splits = 2 if cc_setting in (6, 7) else 0
+    if cc_setting in (3, 4, 5):
+        possible = possible_colors(ct.dark, BITS_PER_COLOR)
+        if len(color_lookups[0]) < len(possible):
+            raise Exception("kabooski")
+        exp = [color for i,color in enumerate(possible) if i in color_lookups[0]] + [(255,255,255)]
+        exp = numpy.array(exp)
+        white = state_info['white']
+        observed = numpy.array([v for k,v in sorted(color_lookups[0].items())] + [white])
+        from colour.characterisation.correction import matrix_colour_correction_Cheung2004
+        der = matrix_colour_correction_Cheung2004(observed, exp)
+
+        # not sure which of this would be better...
+        if ct.ccm is None or cc_setting == 4:
+            ct.ccm = der
+        else:  # cc_setting == 3,5
+            ct.ccm = der.dot(ct.ccm)
+
+    if splits:  # 6,7
```

Comment:
I'd like to come back to this at some point to see if there's something I missed. There's more work to be done here.